core/workflow/safestate_test.go

```diff
@@ -0,0 +1,158 @@
+				}
+				subroles[1] = &taskRole{
+					roleBase: roleBase{state: SafeState{state: task.CONFIGURED}},
+					Traits:   task.Traits{Trigger: "", Await: "", Timeout: "", Critical: true},
+				}
+				ar := &aggregatorRole{
+					roleBase:   roleBase{state: SafeState{state: task.ERROR}},
+					aggregator: aggregator{subroles},
+				}
+				ar.state.merge(task.CONFIGURED, ar)
+				Expect(ar.GetState()).To(Equal(task.CONFIGURED))
+			})
+		})
+		When("the role has at least two sub-roles, state is ERROR and is merged with a healthy state, but other sub-role stays in ERROR", func() {
+			It("should stay in ERROR", func() {
+				subroles := make([]Role, 2)
+				subroles[0] = &taskRole{
+					roleBase: roleBase{state: SafeState{state: task.ERROR}},
+					Traits:   task.Traits{Trigger: "", Await: "", Timeout: "", Critical: true},
+				}
+				subroles[1] = &taskRole{
+					roleBase: roleBase{state: SafeState{state: task.CONFIGURED}},
+					Traits:   task.Traits{Trigger: "", Await: "", Timeout: "", Critical: true},
+				}
+				ar := &aggregatorRole{
+					roleBase:   roleBase{state: SafeState{state: task.ERROR}},
+					aggregator: aggregator{subroles},
+				}
+				ar.state.merge(task.STANDBY, ar)
+				Expect(ar.GetState()).To(Equal(task.ERROR))
+			})
+		})
+		When("the role has at least two sub-roles, the state is healthy and one of non-critical sub-roles enters ERROR", func() {
+			It("should remain in the same state", func() {
+				subroles := make([]Role, 2)
+				subroles[0] = &taskRole{
+					roleBase: roleBase{state: SafeState{state: task.ERROR}},
+					Traits:   task.Traits{Trigger: "", Await: "", Timeout: "", Critical: false},
+				}
+				subroles[1] = &taskRole{
+					roleBase: roleBase{state: SafeState{state: task.CONFIGURED}},
+					Traits:   task.Traits{Trigger: "", Await: "", Timeout: "", Critical: true},
+				}
+				ar := &aggregatorRole{
+					roleBase:   roleBase{state: SafeState{state: task.CONFIGURED}},
+					aggregator: aggregator{subroles},
+				}
+				ar.state.merge(task.ERROR, ar)
+				Expect(ar.GetState()).To(Equal(task.CONFIGURED)) // fixme: this fails
+			})
+		})
+		When("the role has at least two sub-roles, the state is healthy and one of critical sub-roles enters a different state than the other", func() {
+			It("should enter MIXED", func() {
+				subroles := make([]Role, 2)
+				subroles[0] = &taskRole{
+					roleBase: roleBase{state: SafeState{state: task.RUNNING}},
+					Traits:   task.Traits{Trigger: "", Await: "", Timeout: "", Critical: true},
+				}
+				subroles[1] = &taskRole{
+					roleBase: roleBase{state: SafeState{state: task.DONE}},
+					Traits:   task.Traits{Trigger: "", Await: "", Timeout: "", Critical: true},
+				}
+				ar := &aggregatorRole{
+					roleBase:   roleBase{state: SafeState{state: task.RUNNING}},
+					aggregator: aggregator{subroles},
+				}
+				ar.state.merge(task.DONE, ar)
+				Expect(ar.GetState()).To(Equal(task.MIXED))
```

Comment:
This behaves as expected now, but according to the requirements stated in https://its.cern.ch/jira/browse/OCTRL-846, the aggregated state should go to ERROR as long as the expected state is RUNNING. I think we will have to introduce a new concept of an expected state for the tasks...